base/shell/explorer/syspager.cpp

```diff
@@ -178,6 +178,90 @@ class CNotifyToolbar :
     void Initialize(HWND hWndParent, CBalloonQueue * queue);
 };
 
+
+class CHardErrorThread
+{
+    DWORD m_ThreadId;
+    LONG m_bThreadRunning;
+    DWORD m_Status;
+    DWORD m_dwResponseType;
+    CStringW m_Title;
+    CStringW m_Text;
+public:
+
+    CHardErrorThread():
+        m_bThreadRunning(FALSE),
+        m_Status(NULL),
+        m_dwResponseType(NULL),
+        m_ThreadId(NULL)
+    {
+    }
+
+    ~CHardErrorThread()
+    {
+        if (m_bThreadRunning)
+        {
+            /* Try to unstuck Show */
+            PostThreadMessage(m_ThreadId, WM_QUIT, 0, 0);   
+
+            HANDLE hThread = OpenThread(THREAD_ALL_ACCESS, FALSE, m_ThreadId);
+            if (hThread)
+            {
+                WaitForSingleObject(hThread, INFINITE);
+                CloseHandle(hThread);
+            }
+        }
+    }
+
+    VOID ThreadProc()
+    {
+        HRESULT hr;
+        CComPtr<IUserNotification> pnotification;
+
+        OleInitialize(NULL);
+
+        hr = CoCreateInstance(CLSID_UserNotification,
+                              NULL,
+                              CLSCTX_INPROC_SERVER,
+                              IID_PPV_ARG(IUserNotification, &pnotification));
+        if (SUCCEEDED(hr))
+        {
+            hr = pnotification->SetBalloonInfo(m_Title, m_Text, NIIF_WARNING);
+            if (SUCCEEDED(hr))
+            {
+                /* Show will block until the balloon closes */
+                pnotification->Show(NULL, 0);
+            }
+        }
+        InterlockedExchange(&m_bThreadRunning, FALSE);
+    }
+
+    static DWORD CALLBACK s_HardErrorThreadProc(IN OUT LPVOID lpParameter)
+    {
+        reinterpret_cast<CHardErrorThread*>(lpParameter)->ThreadProc();
+        return 0;
+    }
+
+    void StartThread(PBALLOON_HARD_ERROR_DATA pData)
+    {
+        BOOL bIsRunning = InterlockedCompareExchange(&m_bThreadRunning, TRUE, FALSE);
+
+        /* Ignore the new message if we are already showing one */
```

Comment:
Should messages be queued instead, or is this how it's supposed to be?
Not saying you have to do queuing now, just trying to understand for a better perspective on the synchronization stuff.

Reply:
It would be indeed nice to queue them, this would allow the SendMessage called in winsrv to return faster (actually this would also need to be a SendMessageTimeout).

Reply:
windows ignore the message if one is already being shown